Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/StateSynchronization/Observer.cs

```diff
@@ -0,0 +1,228 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Extensions.Experimental.Socketer;
+using System.Collections.Generic;
+using System.IO;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView
+{
+    /// <summary>
+    /// This class observes changes and updates content on a spectator device.
+    /// </summary>
+    public class Observer : Singleton<Observer>
+    {
+        /// <summary>
+        /// Network connection manager that facilitates sending data between devices.
+        /// </summary>
+        [Tooltip("Network connection manager that facilitates sending data between devices.")]
+        [SerializeField]
+        protected TCPConnectionManager connectionManager;
+
+        /// <summary>
+        /// Port used for sending data.
+        /// </summary>
+        [Tooltip("Port used for sending data.")]
+        [SerializeField]
+        protected int port = 7410;
+
+        private SocketEndpoint currentConnection = null;
+        private double[] averageTimePerFeature;
+        private const float heartbeatTimeInterval = 0.1f;
+        private float timeSinceLastHeartbeat = 0.0f;
+        private HologramSynchronizer hologramSynchronizer = new HologramSynchronizer();
+
+        private static readonly byte[] heartbeatMessage = GenerateHeartbeatMessage();
+
+        protected override void Awake()
+        {
+            base.Awake();
+
+            // Ensure that runInBackground is set to true so that the app continues to send network
+            // messages even if it loses focus
+            Application.runInBackground = true;
+
+            if (connectionManager != null)
+            {
+                connectionManager.OnConnected += NetMgr_OnConnected;
+                connectionManager.OnReceive += OnReceive;
+                connectionManager.OnDisconnected += OnDisconnected;
+
+                // Start listening to incoming connections as well.
+                connectionManager.StartListening(port);
+            }
+            else
+            {
+                Debug.LogError("Connection manager not specified for Observer.");
+            }
+        }
+
+        protected void Update()
+        {
+            CheckAndSendHeartbeat();
+            hologramSynchronizer.UpdateHolograms();
+        }
+
+        protected override void OnDestroy()
+        {
+            base.OnDestroy();
+
+            if (connectionManager != null)
+            {
+                connectionManager.DisconnectAll();
```

Comment:
>connectionManager.DisconnectAll(); [](start = 16, length = 34)

We should unsubscribe from the OnConnected, OnReceive, and OnDisconnected events here